mistralrs-core/src/utils/bintokens.rs

```diff
@@ -0,0 +1,181 @@
+                                let s: Vec<char> = s.chars().collect();
+                                if s.len() == 1 {
+                                    space_ch = s[0];
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        if !is_byte_fallback && !is_byte_level {
+            bail!("can't determine decoder type: {:?}", hft.get_decoder());
+        }
+
+        #[allow(clippy::cast_possible_truncation)]
+        let vocab_size = hft.get_vocab_size(true) as u32;
+        let added = hft.get_added_tokens_decoder();
+
+        let mut res = ByteTokenizer {
+            hf_model: "foobar".to_string(),
+            eos_token: 0,
+            vocab_size,
+            special: BTreeMap::new(),
+            token_bytes: (0..vocab_size).map(|_| Vec::new()).collect(),
+            hf_tokenizer: hft,
+        };
+
+        for (id, info) in added.iter() {
+            if info.special {
+                match info.content.as_str() {
+                    "</s>" | "<|endoftext|>" => res.eos_token = *id,
+                    _ => {}
+                }
+                res.special.insert(info.content.clone(), *id);
+            } else {
+                res.token_bytes[*id as usize] = info.content.clone().into_bytes();
+            }
+        }
+
+        let char_map = build_char_map();
+
+        for tok_id in 0..vocab_size {
+            if added.contains_key(&tok_id) {
+                continue;
+            }
+            if let Some(tok_name) = res.hf_tokenizer.id_to_token(tok_id) {
+                if is_byte_fallback {
+                    if tok_name.len() == 6 && tok_name.starts_with("<0x") && tok_name.ends_with('>')
+                    {
+                        // parse hex number from tok_name
+                        let hex_str = &tok_name[3..5];
+                        let byte = u8::from_str_radix(hex_str, 16).unwrap();
+                        res.token_bytes[tok_id as usize] = vec![byte];
+                    } else {
+                        assert!(!tok_name.starts_with("<0x"));
+                        let tok_name = tok_name.replace(space_ch, " ");
+                        res.token_bytes[tok_id as usize] = tok_name.as_bytes().to_vec();
+                    }
+                } else if is_byte_level {
+                    let bytes: Result<Vec<u8>> = tok_name
+                        .chars()
+                        .map(|c| {
+                            char_map
+                                .get(&c)
+                                .copied()
+                                .ok_or_else(|| anyhow!("missing char: {}", c))
+                        })
+                        .collect();
+                    let bytes = match bytes {
+                        Ok(b) => b,
+                        Err(e) => {
+                            println!("error: {} for {:?}", e, tok_name);
```

Comment:
If this is an error, can we propagate it out, or if it is not really an error, log it with `tracing::error!`

Reply:
Sure. The AICI code had some of these, I removed a few but missed this.

The Yacc code has a bunch of these too.

Reply:
Great, thanks!